rust/src/utils/tfrecord.rs

```diff
@@ -0,0 +1,771 @@
+    repeated: bool,
+    feature_type: FeatureType,
+}
+
+impl FeatureMeta {
+    /// Create a new FeatureMeta from a single example.
+    pub fn new(feature: &Feature, is_tensor: bool, is_string: bool) -> Self {
+        let feature_type = match feature.kind.as_ref().unwrap() {
+            Kind::BytesList(data) => {
+                if is_tensor {
+                    let val = &data.value[0];
+                    let tensor_proto = TensorProto::decode(val.as_slice()).unwrap();
+                    FeatureType::Tensor {
+                        shape: tensor_proto
+                            .tensor_shape
+                            .as_ref()
+                            .unwrap()
+                            .dim
+                            .iter()
+                            .map(|d| d.size)
+                            .collect(),
+                        dtype: tensor_proto.dtype(),
+                    }
+                } else if is_string {
+                    FeatureType::String
+                } else {
+                    FeatureType::Binary
+                }
+            }
+            Kind::FloatList(_) => FeatureType::Float,
+            Kind::Int64List(_) => FeatureType::Integer,
+        };
+        Self {
+            repeated: feature_is_repeated(feature),
+            feature_type,
+        }
+    }
+
+    /// Update the FeatureMeta with a new example, or return an error if the
+    /// example is inconsistent with the existing FeatureMeta.
+    pub fn try_update(&mut self, feature: &Feature) -> Result<()> {
+        let feature_type = match feature.kind.as_ref().unwrap() {
+            Kind::BytesList(data) => match self.feature_type {
+                FeatureType::String => FeatureType::String,
+                FeatureType::Binary => FeatureType::Binary,
+                FeatureType::Tensor { .. } => {
+                    let val = &data.value[0];
+                    let tensor_proto = TensorProto::decode(val.as_slice()).unwrap();
+                    FeatureType::Tensor {
+                        shape: tensor_proto
+                            .tensor_shape
+                            .as_ref()
+                            .unwrap()
+                            .dim
+                            .iter()
+                            .map(|d| d.size)
+                            .collect(),
+                        dtype: tensor_proto.dtype(),
+                    }
+                }
+                _ => {
+                    return Err(Error::IO {
+                        message: format!(
+                            "Data type mismatch: expected {:?}, got {:?}",
+                            self.feature_type,
+                            feature.kind.as_ref().unwrap()
+                        ),
+                    })
+                }
+            },
+            Kind::FloatList(_) => FeatureType::Float,
+            Kind::Int64List(_) => FeatureType::Integer,
```

Comment:
This block is very similar to the block in `new`.  Is there a helper method here? `extract_type(feature: &Feature) -> FeatureType`?

Reply:
The full block is subtly different from `new()` (doesn't have access to the same parameters), but I did extract the tensor parsing into a helper, which I think makes it simpler. LMK what you think.